### PR TITLE
feat(taxonomic-filter): batch suggested search results behind a reveal barrier

### PR DESCRIPTION
**Pull request by pauldambra**
## Problem

The Suggested filters tab fired every per-group query in parallel, then revealed each row as its query resolved. Slower groups returning empty pushed earlier rows around as the list reshuffled, and a group with no results above a heavy group could yank the user's hover target out from under them mid-click.

It also showed the same underlying item twice — once as a Recent row (e.g. "Events · pageview") and again as the Events group's own row for the same value — because Recents are computed locally while group top matches arrive separately.

## Changes

- Add a 5s reveal barrier on the Suggested filters tab: on every fresh search, all non-meta group contributions are hidden behind skeletons until every group resolves or the timer fires — whichever comes first. Slow groups keep their skeletons until they complete or error.
- Recents (`name__icontains`) are evaluated client-side and surface immediately — they were already loaded, and revealing them early gives the user something useful while the remote groups settle.
- Dedupe per-group top matches against the recents/pinned rows that are about to render. Build a `${groupType}::${value}` key set from those rows and filter out matching keys from `topMatchItemsWithSkeletons` so the recents row wins (since it's the canonical, user-relevant copy).

## How did you test this code?

I'm an agent (PostHog Code). Automated tests added/run:

- `taxonomicFilterLogic.test.ts`: new "SuggestedFilters dedupe between recents and per-group top matches" describe block — asserts that when a recent's `(sourceGroupType, sourceValue)` matches an item in `topMatchItemsWithSkeletons`, the per-group row is dropped and the recent row survives in `items.results`.
- Existing reveal-barrier and recents tests still pass.
- Manual UI verification done by Paul against the running dev stack.

## Publish to changelog?

no

## 🤖 Agent context

- Authored by PostHog Code (Claude Code) at Paul's direction.
- Iteration order: the reveal barrier landed first; while validating in the running stack we noticed the recents/group duplicate and added the dedupe step.
- Rejected approach: dedupe upstream inside `topMatchItemsWithSkeletons`. We needed access to the rows actually rendered ahead of top matches (recents/pinned, including the no-search-query prefix slices), which live in `infiniteListLogic.items`. Doing it there keeps the dedupe local to the assembly site.
- Per-group top matches keep skeleton placeholders untouched — they still render while the group is loading even if a dedupe key from a recent matches a yet-unseen value.

---
*Created with [PostHog Code](https://posthog.com/code?ref=pr)*

### CODE REVIEW

**Comment by pauldambra** [comment]
* **#59767** <a href="https://app.graphite.com/github/pr/PostHog/posthog/59767?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈 <a href="https://app.graphite.com/github/pr/PostHog/posthog/59767?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* `master`


<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.

File: frontend/src/lib/components/TaxonomicFilter/infiniteListLogic.ts [hunk 1]
```diff
@@ -757,19 +803,66 @@ export const infiniteListLogic = kea<infiniteListLogicType>([
...
+                              }
+                              const group = (item as TaxonomicDefinitionTypes & { group?: TaxonomicFilterGroupType })
+                                  .group
+                              if (!group) {
+                                  return true
+                              }
+                              const sourceGroup = taxonomicGroups.find((g) => g.type === group)
+                              const value = sourceGroup?.getValue?.(item as TaxonomicDefinitionTypes)
+                              if (value == null) {
+                                  return true
```
Comment (pauldambra) [inline frontend/src/lib/components/TaxonomicFilter/infiniteListLogic.ts]: > [!NOTE]
> 🤖 Automated comment by **QA Swarm** — not written by a human

**[code-reviewer]** 🟢 LOW

The dedupe walks `rawTopMatches` and for each non-skeleton calls `taxonomicGroups.find(g => g.type === group)`. Cheap in practice for the small group counts here, but a `Map<groupType, group>` built once before the filter would say the intent more clearly and remove the repeated linear scan.

File: frontend/src/lib/components/TaxonomicFilter/taxonomicFilterLogic.tsx [hunk 2]
```diff
@@ -1892,6 +1932,20 @@ export const taxonomicFilterLogic = kea<taxonomicFilterLogicType>([
...
             const { activeTaxonomicGroup } = values
 
+            // Re-arm the reveal barrier timer on every keystroke. The reducer has already
+            // closed the barrier (when searchQuery is non-empty) — the timer fallback opens
+            // it again after 5s if not all groups have completed. Re-adding under the same
+            // key auto-disposes the previous timer; we only need to explicitly dispose when
+            // the query clears (no fresh timer to schedule).
+            if ((searchQuery ?? '').trim()) {
+                cache.disposables.add(() => {
+                    const timerId = window.setTimeout(() => actions.openRevealBarrier(), 5000)
```
Comment (pauldambra) [inline frontend/src/lib/components/TaxonomicFilter/taxonomicFilterLogic.tsx]: > [!NOTE]
> 🤖 Automated comment by **QA Swarm** — not written by a human

**[code-reviewer]** ⚪ NIT

The 5000ms barrier window is a magic number repeated in two timer call-sites (the afterMount initial-query timer and this setSearchQuery listener) and described as "5s" in comments/tests. Worth a module-level `REVEAL_BARRIER_TIMEOUT_MS = 5000` so the value is named once and changes don't drift.

**Comment by pauldambra** [comment]
> [!NOTE]
> 🤖 Automated comment by **QA Swarm** — not written by a human
>
> Single-reviewer pass (code-reviewer). qa-team specialists and security-audit skipped as not applicable to a focused frontend kea-logic diff.

## Verdict: 💬 APPROVE WITH NITS

Solid, well-tested change. The two-phase reveal (skeletons until barrier opens, recents bypass barrier, dedupe so the same value isn't shown twice) matches the stated UX goal and is covered by both kea-logic unit tests and a `TaxonomicFilter` rendering test.

### Key findings

- 🟢 LOW: `infiniteListLogic.ts` dedupe does an `Array.find` per top-match — a `Map<groupType, group>` built once before the filter would be clearer.
- ⚪ NIT: the `5000` reveal-barrier timeout is duplicated across two timer call-sites and the comments — pull it into a named constant.

### Investigated, not flagged

- The `cache.disposables.add(() => { const t = setTimeout(...); return () => clearTimeout(t) }, key)` pattern looked suspicious at first glance, but `kea-disposables.test.ts` confirms `add(setup, key)` takes a setup factory whose return value is the cleanup — so the timer is correctly scheduled and torn down. No change needed.
- `infiniteListResultsReceived` does `await breakpoint(0)` before checking `anyGroupLoading`. Edge case: a local-only group that synchronously sets `isLoading=false` before remote siblings start could open the barrier prematurely. The existing parameterised test (`zzz-will-not-match` + `event`) exercises Events + Actions which covers this empirically — kept it in mind but not blocking.

### Reviewer summaries

| Reviewer | Assessment |
| --- | --- |
| 🔍 code-reviewer | Two minor improvements worth landing as follow-ups; nothing blocking. |

---
*Automated by QA Swarm — not a human review*

File: frontend/src/lib/components/TaxonomicFilter/infiniteListLogic.ts (hunk 1 above)
Comment (pauldambra) [reply frontend/src/lib/components/TaxonomicFilter/infiniteListLogic.ts]: > [!NOTE]
> 🤖 Automated comment by **PR Shepherd** — not written by a human

Fixed in `8c80526fed3` — `taxonomicGroups` is now built into a `Map<groupType, group>` once per selector evaluation, replacing the per-item `Array.find`.

File: frontend/src/lib/components/TaxonomicFilter/taxonomicFilterLogic.tsx (hunk 2 above)
Comment (pauldambra) [reply frontend/src/lib/components/TaxonomicFilter/taxonomicFilterLogic.tsx]: > [!NOTE]
> 🤖 Automated comment by **PR Shepherd** — not written by a human

Fixed in `8c80526fed3` — added `REVEAL_BARRIER_TIMEOUT_MS = 5000` module constant and replaced both literal call sites.

**Comment by pauldambra** [comment]
> [!NOTE]
> 🤖 Automated comment by **PR Shepherd** — not written by a human

Greptile's parameterisation suggestion applied in `6da44223e87` — collapsed the two recents-surface tests into a single `it.each` matching the skeleton-tests pattern already in the same file.

**Review by adamleithp — APPROVED** [review]
